jujugui/static/gui/src/test/test-bakery.js

```diff
@@ -0,0 +1,256 @@
+
+      bakery.sendRequest('http://example.com/', 'PATCH');
+      assert.equal(client._sendRequest.args[0][0], 'patch');
+      client._sendRequest.reset();
+    });
+
+    it('sets the headers', () => {
+      bakery.sendRequest('http://example.com/', 'GET', {'foo': 'bar'});
+      const expectedHeaders = {
+        'Bakery-Protocol-Version': 1,
+        'foo': 'bar'
+      };
+      assert.deepEqual(client._sendRequest.args[0][2], expectedHeaders);
+    });
+
+    it('adds macaroons to the request', () => {
+      // We add two "macaroons" into the store--one for the url we're setting,
+      // one that we should not get.
+      fakeLocalStorage.store['charmstore'] = 'doctor';
+      fakeLocalStorage.store['identity'] = 'bad wolf';
+
+      bakery.sendRequest('http://example.com/', 'GET');
+      const expectedHeaders = {
+        'Bakery-Protocol-Version': 1,
+        'Macaroons': 'doctor'
+      };
+      assert.deepEqual(client._sendRequest.args[0][2], expectedHeaders);
+    });
+
+    it('wraps callbacks with discharge functionality', () => {
+      const wrapper = sinon.stub(bakery, '_wrapCallback');
+      bakery.sendRequest('http://example.com/', 'GET');
+      assert.equal(wrapper.callCount, 1);
+    });
+  });
+
+  describe('macaroon discharges', () => {
+    let dischargeStub, importStub, exportStub;
+
+    beforeEach(() => {
+      dischargeStub = sinon.stub(macaroonlib, 'discharge');
+      importStub = sinon.stub(macaroonlib, 'import');
+      exportStub = sinon.stub(macaroonlib, 'export').returnsArg(0);
+    });
+
+    afterEach(() => {
+      dischargeStub.restore();
+      importStub.restore();
+      exportStub.restore();
+    });
+
+    it('discharges macaroons', () => {
+      const macaroon = 'this is a macaroon';
+      dischargeStub.callsArgWith(2, macaroon);
+      const success = discharges => {
+        assert.equal(discharges, macaroon);
+      };
+      const failure = msg => {
+        console.error(msg);
+        assert.fail();
+      };
+      bakery.discharge('macaroons', success, failure);
+      assert.equal(
+        dischargeStub.callCount, 1,
+        'macaroonlib discharge not called');
+      assert.equal(importStub.callCount, 1, 'macaroonlib import not called');
+      assert.equal(exportStub.callCount, 1, 'macaroonlib export not called');
+    });
+
+    it('handles failures discharging macaroons', () => {
+      const macaroon = 'this is a macaroon';
+      const error = { message: 'bad wolf' };
+      dischargeStub.callsArgWith(2, macaroon);
+      importStub.throws(error);
+      const success = () => {};
+      const failure = msg => {
+        assert.equal(msg, `discharge failed: ${error.message}`);
+      };
+      bakery.discharge('macaroons', success, failure);
+    });
+
+    it('handles third party discharge', () => {
+      const condition = 'this is a caveat';
+      const success = macaroons => {}; 
+      const failure = err => {};
+      bakery._getThirdPartyDischarge(
```

Comment:
Is it possible to test this only using the public API? So basically passing a macaroon with a third party?
Anyway, thanks for these tests, it feels like we are already exercising much more than before.